ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,316 @@
+        pos_control->set_alt_target_to_current_alt();
+        pos_control->set_desired_velocity_z(inertial_nav.get_velocity_z());
+    }
+
+    // initialise waypoint state
+    zigzag_is_between_A_and_B = false;
+    zigzag_judge_moving.is_keeping_time = false;
+    stage = REQUIRE_A;     
+    return true;   
+}
+
+// run - runs the zigzag controller
+// should be called at 100hz or more
+void Copter::ModeZigZag::run()
+{
+    // if not auto armed or motors not enabled set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || !motors->get_interlock() || ap.land_complete) {
+        zero_throttle_and_relax_ac();
+        return;
+    }
+
+    // manual control activated when point A B is not defined
+    if (stage == REQUIRE_A || stage == REQUIRE_B || stage == MANUAL_REGAIN) {
+        // receive pilot's inputs, do position and attitude control
+        manual_control();
+    } else { 
+        //auto flight
+        //judge if the vehicle has arrived at the current destination
+        //if yes, go to the manual control stage
+        //else, fly to current destination
+        if (has_arr_at_dest()) {  //if the vehicle has arrived at the current destination
+            stage = MANUAL_REGAIN;
+            loiter_nav->init_target();
+        } else {
+            auto_control();
+        }
+    }
+}
+
+//auto_control - guide the vehicle to fly to current destination
+void Copter::ModeZigZag::auto_control()
+{
+    // process pilot's yaw input
+    float target_yaw_rate = 0;
+    if (!copter.failsafe.radio) {
+        // get pilot's desired yaw rate
+        target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+    }
+
+    // set motors to full range
+    motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
+
+    // run waypoint controller to update xy
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+
+    // call z-axis position controller (wpnav should have already updated it's alt target)
+    pos_control->update_z_controller();
+
+    // call attitude controller
+    // roll & pitch from waypoint controller, yaw rate from pilot
+    attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), target_yaw_rate);        
+}
+
+//manual_control - process manual control
+void Copter::ModeZigZag::manual_control()
+{
+    float target_yaw_rate = 0.0f;
+    float target_climb_rate = 0.0f;
+
+    // initialize vertical speed and acceleration's range
+    pos_control->set_speed_z(-get_pilot_speed_dn(), g.pilot_speed_up);
+    pos_control->set_accel_z(g.pilot_accel_z);
+    // process pilot inputs unless we are in radio failsafe
+    if (!copter.failsafe.radio) {
+        // apply SIMPLE mode transform to pilot inputs
+        update_simple_mode();
+        // process pilot's roll and pitch input
+        loiter_nav->set_pilot_desired_acceleration(channel_roll->get_control_in(), channel_pitch->get_control_in(), G_Dt);
```

Comment:
I think this is incorrect.  We are passing the pilot's input (a number from -4500 to +4500) into this function which accepts desired accelerations.  I think we should be making a call to get_pilot_desired_lean_angles() like we do in Loiter mode.